File: 1329-sort-the-matrix-diagonally/1329-sort-the-matrix-diagonally.java
class Solution {
    public int[][] diagonalSort(int[][] mat) {
      int m=mat.length,n=mat[0].length;
      //row=0
        for(int col=0;col<n;col++){
            sort(mat,0,col,m,n);
        }
        //col=0
        for(int row=0;row<m;row++){
            sort(mat,row,0,m,n);
        }
        return mat;
    }
    public static void sort(int[][] mat,int row,int col,int m,int n){
        int[] values=new int[101];//using count sort
        int r=row,c=col;
        while(r<m&&c<n){
            values[mat[r][c]]++;
            r++;
            c++;
        }
        r=row;
        c=col;
        for(int i=1;i<101;i++){
            if(values[i]>0){
                int cnt=values[i];
                while(cnt-->0){
                    mat[r][c]=i;
                    r++;c++;
                }
            }
        }
    }
}

File: 1329-sort-the-matrix-diagonally/1329-sort-the-matrix-diagonally.cpp
class Solution {
public:
    vector<vector<int>> diagonalSort(vector<vector<int>>& mat) { 

// // Explanation
// // A[i][j] on the same diagonal have same value of i - j
// // For each diagonal,
// // put its elements together, sort, and set them back.


// // Complexity
// // Time O(MNlogD), where D is the length of diagonal with D = min(M,N).
// // Space O(MN)

        // solving with priority queue, 20ms
//         int n = mat.size(), m = mat[0].size();
//         unordered_map<int, priority_queue<int, vector<int>, greater<int>>> mp;
//         for(int i=0;i<n;i++){
//             for(int j=0;j<m;j++){
//                 mp[i-j].push(mat[i][j]);
//             }
//         }   
//         for(int i=0;i<n;i++){
//             for(int j=0;j<m;j++){
//                 mat[i][j] = mp[i-j].top();
//                 mp[i-j].pop();
//             }
//         }
//         return mat;



        // 2. solving with sorting techniques, 8ms 
        int r = mat.size(), c = mat[0].size();
        int j,p,q;
        // solving upper right triangle
        for(int i=0;i<c;i++){
            // extracting the values 
            p=0,q=i;
            vector<int> t;
            while(p < r and q < c){
                t.push_back(mat[p][q]);
                p++;q++;
            }
            sort(t.begin(), t.end());
            // placing back
            p=0,q=i,j=0;
            while(p<r and q<c){
                mat[p][q] = t[j];
                p++;q++;j++;
            }
        }

        // solving lower left triangle
        for(int i=1;i<r;i++){
            // extracting the values
            vector<int> t;
            p=i,q=0;
            while(p<r and q<c){
                t.push_back(mat[p][q]);
                p++;q++;
            }
            sort(t.begin(),t.end());
            // placing back
            p=i,q=0,j=0;
            while(p<r and q<c){
                mat[p][q] = t[j];
                p++;q++;j++;
            }
        }
        return mat;
    }
};
